### PR TITLE
Add flag `set-env-var` to `MixEnvironment`, and rename old ones

### PR DESCRIPTION
**Pull request by bryanhonof**
# Motivation

Closes: #11495 
Closes: #10338

# Context

https://github.com/NixOS/nix/pull/11490#discussion_r1757055057

Ref: #11490

# Priorities and Process

Add :+1: to [pull requests you find important](https://github.com/NixOS/nix/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc).

The Nix maintainer team uses a [GitHub project board](https://github.com/orgs/NixOS/projects/19) to [schedule and track reviews](https://github.com/NixOS/nix/tree/master/maintainers#project-board-protocol).

### CODE REVIEW

**Review by roberth — CHANGES_REQUESTED** [review]
I like that this makes the flag names more consistent, but I'm concerned about the user impact, since many users have opted in to the `nix-command` feature flag. Perhaps the old names could be kept?

Another issue is that the effect of `--set-env-var FOO` followed by `--unset-env-var FOO` may be unexpected. These should be either mutually inclusive, or handled in the order they occur, e.g. by using a `map<string, optional<string>>` to represent both setting and unsetting/keeping in a single map value. Using `nullopt` for both unsetting and keeping might be a little weird, so you could also do `map<string, variant<bool, string>>` instead.

This definitely needs new tests cases to
- test the new flag
- test the interactions.

**Comment by bryanhonof** [comment]
I have another question following yours.
What if the user does the following?

```console
$ nix run --set-env-var FOO  '' --unset-env-var FOO --set-env-var FOO ''...
```

Since we're talking about storing it all in a `map`, the second appearance of `--set-env-var FOO` will just be ignored altogether, right?
I don't believe it'd be correct to add "weight" to certain flags, so they do survive something like this, maybe throw a warning to the user they've added a duplicate flag?

**Comment by bryanhonof** [comment]
I gave the tests a go, let me know what you think.

**Comment by bryanhonof** [comment]
Apparently, also fixes one more shellcheck case from #10795.

File: tests/functional/flakes/develop.sh [hunk 1]
```diff
@@ -43,13 +43,92 @@ echo "\$ENVVAR"
...
+(
+  expect=''
+  got="$(nix develop --set-env-var FOO 'BAR' --unset-env-var FOO --no-write-lock-file .#hello <<EOF
+echo "\$FOO"
+EOF
+)"
+  [[ "$got" == "$expect" ]]
+)
+
+# Test that `--unset-env-var` unsets the variable, regardless of where in the command it is.
```
Comment (roberth) [inline tests/functional/flakes/develop.sh:83]: I think this should be an error, as it is quite likely that the user made a mistake by both setting and unsetting a variable.
Changing it to an error may also let us change the semantics to first-wins or last-wins in the future.

File: tests/functional/flakes/develop.sh [hunk 2]
```diff
@@ -43,13 +43,92 @@ echo "\$ENVVAR"
...
+echo "\$FOO"
+EOF
+)"
+  [[ "$got" == "$expect" ]]
+)
+
+# Test that `--set-env-var` overwrites previously set variables.
+(
+  expect='BLA'
+  got="$(nix develop --set-env-var FOO 'BAR' --set-env-var FOO 'BLA' --no-write-lock-file .#hello <<EOF
```
Comment (roberth) [inline tests/functional/flakes/develop.sh:96]: I have to inform you that IBM used _BLA_ in all seriousness for Business Layer Application in one of their products. I will read it as such.

File: src/libcmd/command.cc [hunk 3]
```diff
@@ -282,51 +284,78 @@ MixDefaultProfile::MixDefaultProfile()
...
+    : ignoreEnvironment(false)
 {
     addFlag({
-        .longName = "ignore-environment",
+        .longName = "ignore-env",
+        .aliases = {"ignore-environment"},
         .shortName = 'i',
-        .description = "Clear the entire environment (except those specified with `--keep`).",
+        .description = "Clear the entire environment, except for those specified with `--keep-env-var`. "
+                       "Alias for `--ignore-environment`.",
```
Comment (roberth) [inline src/libcmd/command.cc:295]: This suggests that `--ignore-environment` is what we should use. I thought the goal was to standardize on the `env` variations.

File: src/libcmd/command.cc [hunk 4]
```diff
@@ -282,51 +284,78 @@ MixDefaultProfile::MixDefaultProfile()
...
-            throw UsageError("--keep does not make sense without --ignore-environment");
+    auto env = getEnv();
 
-        for (const auto & var : unset)
-            unsetenv(var.c_str());
-    }
+    if (ignoreEnvironment)
+        std::erase_if(env, [&](const auto & var) { return !keepVars.contains(var.first); });
+
+    if (!setVars.empty())
```
Comment (roberth) [inline src/libcmd/command.cc:349]:
```suggestion
```

Equivalent

File: tests/functional/flakes/develop.sh (hunk 1 above)
Comment (roberth) [reply tests/functional/flakes/develop.sh:83]: We also have a similar ambiguity in the `--ignore-environment` `--keep-var` situation as well.

Comment (bryanhonof) [reply tests/functional/flakes/develop.sh:83]: For the `--ignore-environment` `--keep-var` case, we have this:

```cpp
    if (!ignoreEnvironment && !keepVars.empty())
        throw UsageError("--keep-env-var does not make sense without --ignore-env");
```

File: tests/functional/flakes/develop.sh (hunk 2 above)
Comment (bryanhonof) [reply tests/functional/flakes/develop.sh]: Good to know. 😁

**Review by roberth — APPROVED** [review]

**Pull request merged**